dot15d4/src/csma/mod.rs

```diff
@@ -1,12 +1,909 @@
+    ignore_not_for_us: bool,
+    /// Even if there is no ack_request flag set, ack it anyway
+    ack_everything: bool,
+}
+
+impl Default for CsmaConfig {
+    fn default() -> Self {
+        Self {
+            ack_unicast: true,
+            ack_broadcast: false,
+            ignore_not_for_us: true,
+            ack_everything: false,
+        }
+    }
+}
+
+/// Structure that setups the CSMA futures
+pub struct CsmaDevice<R: Radio, Rng, D: Driver, TIMER> {
+    radio: Mutex<R>,
+    rng: Mutex<Rng>,
+    driver: D,
+    timer: TIMER,
+    hardware_address: [u8; 8],
+    config: CsmaConfig,
+}
+
+impl<R, Rng, D, TIMER> CsmaDevice<R, Rng, D, TIMER>
+where
+    R: Radio,
+    Rng: RngCore,
+    D: Driver,
+{
+    /// Creates a new CSMA object that is ready to be run
+    pub fn new(radio: R, rng: Rng, driver: D, timer: TIMER, config: CsmaConfig) -> Self {
+        let hardware_address = radio.ieee802154_address();
+        CsmaDevice {
+            radio: Mutex::new(radio),
+            rng: Mutex::new(rng),
+            driver,
+            timer,
+            hardware_address,
+            config,
+        }
+    }
+}
+
+impl<R, Rng, D, TIMER> CsmaDevice<R, Rng, D, TIMER>
+where
+    R: Radio,
+    for<'a> R::RadioFrame<&'a mut [u8]>: RadioFrameMut<&'a mut [u8]>,
+    for<'a> R::TxToken<'a>: From<&'a mut [u8]>,
+    Rng: RngCore,
+    D: Driver,
+    TIMER: DelayNs + Clone,
+{
+    /// Run the CSMA module. This should be run in its own task and polled seperately.
+    pub async fn run(&mut self) -> ! {
+        let mut wants_to_transmit_signal = Channel::new();
+        let (sender, receiver) = wants_to_transmit_signal.split();
+        self.radio.get_mut().enable().await; // Wake up radio
+        match select::select(
+            self.transmit_package_task(sender),
+            self.receive_package_task(receiver),
+        )
+        .await
+        {
+            Either::First(_) => panic!("Tasks should never terminate, csma transmission just did"),
+            Either::Second(_) => panic!("Tasks should never terminate, csma receiving just did"),
+        }
+    }
+
+    /// Checks if the current frame is intended for us. For the hardware address, the full 64bit
+    /// address should be provided.
+    fn is_package_for_us<BUF: AsRef<[u8]>>(hardware_address: &[u8; 8], frame: &Frame<BUF>) -> bool {
+        let Some(addr) = frame
+            .addressing()
+            .and_then(|fields| fields.dst_address(&frame.frame_control()))
+        else {
+            return false;
+        };
+
+        match &addr {
+            _ if addr.is_broadcast() => true,
+            Address::Absent => false,
+            Address::Short(addr) => hardware_address[6..] == addr[..2],
+            Address::Extended(addr) => hardware_address == addr,
+        }
+    }
+
+    async fn receive_package_task(&self, mut wants_to_transmit_signal: Receiver<'_, ()>) -> ! {
```

Comment:
`receive_package_task` -> `receive_frame_task`.

Could you also add documentation to the function, even for non-public functions?